### ISSUE TITLE
Unable to use Object type within select expression with if-else expression

### ISSUE THREAD

**Issue by mistyharsh** [edgedb/edgedb#1593]
- EdgeDB Version: 1.0 Alpha 4
- OS Version: Docker

Steps to Reproduce:

1. Create a database with simple Object type.
2. Use `if-else` expression within the function definition.

Schema:

```edgeql
type Project {
  required property name -> str;
};

create function add_project(n1: str, n2: str) -> Project {
  set volatility := 'VOLATILE';
  using EdgeQL $$
    with newItem1 := (insert default::Project {
      name := n1
    }),
    newItem2 := (insert default::Project {
      name := n2
    })
    select newItem1 if n1 = 'hello' else newItem2
  $$;
};
```

The error is:
```
ERROR: InternalServerError: could not serialize result in worker subprocess
  Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
  Server traceback:
      cannot pickle 'weakref' object:

      Traceback (most recent call last):
        File "/usr/lib/x86_64-linux-gnu/edgedb-server-1-alpha4/lib/python3.8/site-packages/edb/server/procpool/worker.py", line 88, in worker
          pickled = pickle.dumps(data)
      TypeError: cannot pickle 'weakref' object
```

Creation of function `add_project` fails.

**Comment by elprans** [comment]
Couple of issues here.

First, the error propagation crash is obviously a bug.

Second is that EdgeQL queries, like SQL they compile to, are not imperative scripts, and, in general, you cannot make any assumptions about the execution flow (see the [relevant section](https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-EXPRESS-EVAL) in Postgres docs).  So in the following

```
WITH 
    A := (INSERT ...),
    B := (INSERT ...)
SELECT
    A IF condition ELSE B
```

you cannot assume that either `A` or `B` will not be computed simply because they are referred to in a subsequent `IF`.  In fact, in the query above, EdgeDB will actually assume that you want to execute _both_ `INSERT` statements and then return the result of either depending on the condition.  And if you put `INSERT` directly into an `IF` statement, EdgeDB will return an error, because then the conditional intent is clear and there's no generic way to realize it in SQL ([this is why](https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-MODIFYING)).

We are considering adding limited support for `INSERT` as the right-hand operand in `??` to support the very common "select-or-insert" and "update-or-insert" patterns which would rely on constraints (see #684 and #1467 for more discussion about this).

Finally, we need to do a better job at documenting these caveats clearly.  \cc @vpetrovykh

**Comment by mistyharsh** [comment]
Thanks @elprans for the explanation. That makes sense now. However, I have a follow-up question and rather a dilemma. I am using EdgeDB with .NET Core. Since, the driver doesn't exist, I rely on HTTP protocol to issue DB queries which obviously presents a transaction problem. As you mentioned in another [thread](https://github.com/edgedb/edgedb/discussions/1583#discussioncomment-42271), the goal should be to minimize and use atomic queries as much as possible, how should I go about using complex/mixed queries like `delete` followed by `insert` or `delete` followed by `update` commands?

I understand that **EdgePL** is planned post v1 and it should enable such cases by the virtue of imperative control flow statements, but is there any workaround to address this today?

**Comment by elprans** [comment]
> is there any workaround to address this today?

One way would be to allow sending EdgeQL scripts (i.e. multiple statements) in HTTP requests.  That way you'd be able to execute your entire transaction body in a single request.  That would become especially useful once we implement support for arguments in scripts (edgedb/edgedb-python#51).

**Comment by mistyharsh** [comment]
Thanks @elprans. For now, I have found out a temporary alternative. From our `.NET` app, we will spin off a separate `Node.js` process and use JS driver to connect to DB and connect over HTTP. We will implement simple timestamp-based transaction.

I hope that soon we would have driver for `.NET Core`. I tried writing the driver but could only do basic handshake. Full-fledged low-level communication is not my cup of tea.
